---
title: A few small simulations to assess benefits of NIG prior in various settings
author: Peter Carbonetto
output: workflowr::wflow_html
---

Here I summarize some of the results of my "small" simulations to
compare the original SuSiE model to SuSiE with an NIG prior (also
called the "Servin and Stephens", or "SS", prior). In the following,
we use "SuSiE-SS" as a shorthand for this variant of SuSiE.

```{r knitr-opts, include=FALSE}
knitr::opts_chunk$set(comment = "#",collapse = TRUE,results = "hold",
                      fig.align = "center",dpi = 120)
```

Load the packages needed to perform the analyses below.

```{r load-pkgs, message=FALSE}
library(ggplot2)
library(cowplot)
```

Set the seed for reproducibility:

```{r set-seed}
set.seed(1)
```

This is a function I will use to empirically assess power and coverage
from the simulations.

```{r define-compute-power-and-coverage}
compute_power_and_coverage <- function (res, causal) {
  N        <- length(res)
  power    <- 0
  coverage <- 0
  V_true  <- NULL
  V_false <- NULL
  for (i in 1:N) {
    V        <- res[[i]]$V
	cs       <- res[[i]]$sets$cs
	L        <- length(V)
	all_cs   <- paste0("L",seq(1,L))
    names(V) <- all_cs
    if (length(cs) == 0) {
      x <- NULL
	  y <- NULL
    } else {
      x <- intersect(unique(unlist(cs)),causal[[i]])
	  y <- names(which(sapply(cs, 
             function (x) length(intersect(causal_snps[[i]],x)) > 0)))
    }
    V_true   <- c(V_true,V[y])
    V_false  <- c(V_false,V[setdiff(all_cs,y)])
    power    <- power    + length(x)
	coverage <- coverage + length(y)
  }
  num_true <- sum(sapply(causal,length))
  num_pos  <- sum(sapply(res,function (x) length(x$sets$cs)))
  return(list(power    = power/num_true,
              coverage = coverage/num_pos,
              V_true   = V_true,
			  V_false  = V_false))
}
```

And this a function I will use to compute the sizes of the credible
sets:

```{r define-get-cs-sizes}
get_cs_sizes <- function (res)
  unlist(lapply(res,function (x) sapply(x$sets$cs,length)))
```

## *n* = 500

In this setting, the sample size is large enough ($n = 500$) that I
do not expect to see much benefit of using the NIG prior.

```{r load-output-n-500}
load("../output/small_sim_out_n=500_v0.14.47.RData")
```

Compare power and coverage:

```{r power-coverage-n-500}
stats_susie    <- compute_power_and_coverage(res_susie,causal_snps)
stats_susie_ss <- compute_power_and_coverage(res_susie_small,causal_snps)
data.frame(method   = c("susie","susie_ss"),
           power    = c(stats_susie$power,stats_susie_ss$power),
  	       coverage = c(stats_susie$coverage,stats_susie_ss$coverage))
```

These are the running times:

```{r runtimes-n-500}
summary(runtimes)
```

These are the CS sizes:

```{r cs-sizes-n-500, fig.height=3.5, fig.width=3}
sizes_susie    <- get_cs_sizes(res_susie)
sizes_susie_ss <- get_cs_sizes(res_susie_small)
compare_cs_sizes <- function (susie, susie_ss, max_size = Inf) {
  pdat <- rbind(data.frame(method = "susie",   size = susie),
                data.frame(method = "susie_ss",size = susie_ss))
  pdat <- subset(pdat,size <= max_size)
  return(ggplot(pdat,aes(x = size,color = method,fill = method)) +
         facet_grid(rows = vars(method),scales = "free_y",axes = "all") +
		 geom_histogram(bins = 24) +
		 scale_color_manual(values = c("darkblue","darkorange")) +
		 scale_fill_manual(values = c("darkblue","darkorange")) +
		 labs(x = "CS size") +
		 theme_cowplot(font_size = 10) +
		 theme(strip.background = element_blank()))
}
compare_cs_sizes(sizes_susie,sizes_susie_ss,max_size = 20)
```

Compare the prior variances of the true positives vs. the false
positives:

```{r V-n-500, fig.height=2, fig.width=4.25}
compare_prior_variances <- function (susie_true, susie_false,
                                     susie_ss_true, susie_ss_false,
									 ymax = Inf) {
  pdat <- rbind(data.frame(method="susie",causal=TRUE,V=susie_true),
                data.frame(method="susie",causal=FALSE,V=susie_false),
                data.frame(method="susie_ss",causal=TRUE,V=susie_ss_true),
                data.frame(method="susie_ss",causal=FALSE,V=susie_ss_false))
  pdat <- transform(pdat,sigma = sqrt(V))
  p <- ggplot(pdat,aes(x = sigma,color = causal,fill = causal)) +
    facet_grid(cols = vars(method),scales = "free_y",axes = "all") +
    geom_histogram(bins = 18,position = "stack",linewidth = 0.05) +
	scale_color_manual(values = c("darkblue","orangered")) +
	scale_fill_manual(values = c("darkblue","orangered")) +
	labs(x = "prior s.d.") +
	theme_cowplot(font_size = 10) +
	theme(strip.background = element_blank())
  if (is.finite(ymax))
    p <- p + coord_cartesian(ylim = c(0,ymax))
  return(p)
}
compare_prior_variances(stats_susie$V_true,stats_susie$V_false,
                        stats_susie_ss$V_true,stats_susie_ss$V_false,
						ymax = 75)
```

Note that most of the prior variances are at or near zero; the Y axis
was cut off at 75 so that we can see the larger prior variances more
clearly.

## *n* = 200

Next let's consider a setting with slightly smaller sample size.

```{r load-output-n-200}
load("../output/small_sim_out_n=200_v0.14.47.RData")
```

Notice that SuSiE-SS substantially improves power and coverage only
diminishes slightly:

```{r power-coverage-n-200}
stats_susie    <- compute_power_and_coverage(res_susie,causal_snps)
stats_susie_ss <- compute_power_and_coverage(res_susie_small,causal_snps)
data.frame(method   = c("susie","susie_ss"),
           power    = c(stats_susie$power,stats_susie_ss$power),
  	       coverage = c(stats_susie$coverage,stats_susie_ss$coverage))
```

These are the running times:

```{r runtimes-n-200}
summary(runtimes)
```

These are the CS sizes:

```{r cs-sizes-n-200, fig.height=3.5, fig.width=3}
sizes_susie    <- get_cs_sizes(res_susie)
sizes_susie_ss <- get_cs_sizes(res_susie_small)
compare_cs_sizes(sizes_susie,sizes_susie_ss,max_size = 40)
```

Compare the prior variances of the true positives vs. the false
positives:

```{r V-n-200, fig.height=2, fig.width=4.25}
compare_prior_variances(stats_susie$V_true,stats_susie$V_false,
                        stats_susie_ss$V_true,stats_susie_ss$V_false,
						ymax = 250)
```

## *n* = 200, MAF < 5%

If we focus on fine-mapping causal SNPs with low minor allele
frequencies (MAF < 5%), something interesting happens. Overall, we
have reduced coverage—and this may be because we are assuming at least
one "significant" association at the locus—nonetheless SuSiE-SS now
provides a greater gain in power over SuSiE:

```{r power-coverage-n-200-maf-0.05}
load("../output/small_sim_out_n=100_maf=0.05_v0.14.47.RData")
stats_susie    <- compute_power_and_coverage(res_susie,causal_snps)
stats_susie_ss <- compute_power_and_coverage(res_susie_small,causal_snps)
data.frame(method   = c("susie","susie_ss"),
           power    = c(stats_susie$power,stats_susie_ss$power),
  	       coverage = c(stats_susie$coverage,stats_susie_ss$coverage))
```

These are the running times:

```{r runtimes-n-200-maf-0.05}
summary(runtimes)
```

These are the CS sizes:

```{r cs-sizes-n-200-maf-0.05, fig.height=3.5, fig.width=3}
sizes_susie    <- get_cs_sizes(res_susie)
sizes_susie_ss <- get_cs_sizes(res_susie_small)
compare_cs_sizes(sizes_susie,sizes_susie_ss,max_size = 40)
```

Compare the prior variances of the true positives vs. the false positives:

```{r V-n-200-maf-0.05, fig.height=2, fig.width=4.25}
compare_prior_variances(stats_susie$V_true,stats_susie$V_false,
                        stats_susie_ss$V_true,stats_susie_ss$V_false,
						ymax = 150)
```

## *n* = 40

At a much smaller sample size, the advantages of SuSiE-SS become even
more clear.

```{r power-coverage-n-40}
load("../output/small_sim_out_n=40_v0.14.47.RData")
stats_susie    <- compute_power_and_coverage(res_susie,causal_snps)
stats_susie_ss <- compute_power_and_coverage(res_susie_small,causal_snps)
data.frame(method   = c("susie","susie_ss"),
           power    = c(stats_susie$power,stats_susie_ss$power),
  	       coverage = c(stats_susie$coverage,stats_susie_ss$coverage))
```

These are the running times:

```{r runtimes-n-40}
summary(runtimes)
```

These are the CS sizes:

```{r cs-sizes-n-40, fig.height=3.5, fig.width=3}
sizes_susie    <- get_cs_sizes(res_susie)
sizes_susie_ss <- get_cs_sizes(res_susie_small)
compare_cs_sizes(sizes_susie,sizes_susie_ss,max_size = 40)
```

Compare the prior variances of the true positives vs. the false
positives:

```{r V-n-40, fig.height=2, fig.width=4.25}
compare_prior_variances(stats_susie$V_true,stats_susie$V_false,
                        stats_susie_ss$V_true,stats_susie_ss$V_false,
						ymax = 200)
```
